basenode/src/main/java/io/coti/basenode/data/ExecutorData.java

```diff
@@ -1,73 +1,85 @@
 import lombok.Data;
 import lombok.extern.slf4j.Slf4j;
 
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
-import java.util.concurrent.ExecutionException;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
-import java.util.concurrent.Future;
-import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
 
 @Slf4j
 @Data
 public class ExecutorData {
 
     private final ExecutorService executorService;
-    private final List<Future<?>> futures;
     private final InitializationTransactionHandlerType initializationTransactionHandlerType;
+    private AtomicLong completedFutures;
+    private AtomicLong failedFutures;
+    private AtomicLong transactionsToProcess;
 
-    public ExecutorData(InitializationTransactionHandlerType initializationTransactionHandlerType) {
+    public ExecutorData(InitializationTransactionHandlerType initializationTransactionHandlerType, AtomicLong transactionsToProcess) {
         executorService = Executors.newFixedThreadPool(1);
-        futures = Collections.synchronizedList(new ArrayList<>());
         this.initializationTransactionHandlerType = initializationTransactionHandlerType;
+        completedFutures = new AtomicLong(0);
+        failedFutures = new AtomicLong(0);
+        this.transactionsToProcess = transactionsToProcess;
     }
 
     public void submit(Runnable runnable) {
-        futures.add(executorService.submit(runnable));
+        CompletableFuture.runAsync(runnable, executorService)
+                .exceptionally(exception -> {
+                    log.error("Error while running " + initializationTransactionHandlerType + " executor.\n", exception);
+                    failedFutures.incrementAndGet();
+                    return null;
+                }).thenRunAsync(() -> completedFutures.incrementAndGet());
     }
 
     public void waitForTermination() {
-        AtomicInteger completedFutures = new AtomicInteger(0);
-        Thread monitorCompletionThread = getMonitorCompletionThread(completedFutures);
+        Thread monitorCompletionThread = getMonitorCompletionThread();
         monitorCompletionThread.start();
-        for (Future<?> future : futures) {
-            try {
-                future.get();
-                completedFutures.incrementAndGet();
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-            } catch (ExecutionException e) {
-                log.error(e.getMessage());
-            }
-        }
-        executorService.shutdown();
-        monitorCompletionThread.interrupt();
+        int completedCheck = 0;
+        AtomicLong completedFuturesCheck = completedFutures;
         try {
+            while (completedFutures.get() != transactionsToProcess.get()) {
+                Thread.sleep(100);
+                completedCheck++;
```

Comment:
the entire block of check should refactor to a seperate method of futureTimeout return boolean if passed 300

Reply:
psudo:
 while (completedFutures.get() != transactionsToProcess.get()  && ! futureWaitTimedOut()  ) {